packages/opencensus-instrumentation-grpc/src/grpc.ts

```diff
@@ -0,0 +1,478 @@
+        // span.
+        if (!plugin.tracer.currentRootSpan) {
+          return plugin.tracer.startRootSpan(
+              traceOptions,
+              plugin.makeGrpcClientRemoteCall(original, args, this, plugin));
+        } else {
+          const span = plugin.tracer.startChildSpan(
+              traceOptions.name, traceOptions.type);
+          return (plugin.makeGrpcClientRemoteCall(
+              original, args, this, plugin))(span);
+        }
+      };
+    };
+  }
+
+  /**
+   * This method handels the client remote call
+   */
+  private makeGrpcClientRemoteCall(
+      // tslint:disable-next-line:no-any
+      original: any, args: any[], self: any, plugin: GrpcPlugin) {
+    /**
+     * Patches a callback so that the current span for this trace is also ended
+     * when the callback is invoked.
+     */
+    function patchedCallback(
+        span: types.Span,
+        // tslint:disable-next-line:no-any
+        callback: any) {
+      // tslint:disable-next-line:no-any
+      const wrappedFn = (err: grpcModule.ServiceError, res: any) => {
+        if (err) {
+          span.status = plugin.traceStatus(err.code);
+          span.addAttribute(
+              GrpcPlugin.ATTRIBUTE_GRPC_STATUS_CODE, err.code.toString());
+          span.addAttribute(GrpcPlugin.ATTRIBUTE_GRPC_ERROR_NAME, err.name);
+          span.addAttribute(
+              GrpcPlugin.ATTRIBUTE_GRPC_ERROR_MESSAGE, err.message);
+        } else {
+          span.status = plugin.traceStatus(grpcModule.status.OK);
+          span.addAttribute(
+              GrpcPlugin.ATTRIBUTE_GRPC_STATUS_CODE,
+              grpcModule.status.OK.toString());
+        }
+        span.end();
+        callback(err, res);
+      };
+      return plugin.tracer.wrap(wrappedFn);
+    }
+
+    return (span: types.Span) => {
+      if (!span) {
+        return original.apply(self, args);
+      }
+
+      // if unary or clientStream
+      if (!original.responseStream) {
+        const callbackFuncIndex = findIndex(args, (arg) => {
+          return typeof arg === 'function';
+        });
+        if (callbackFuncIndex !== -1) {
+          args[callbackFuncIndex] =
+              patchedCallback(span, args[callbackFuncIndex]);
+        }
+      }
+
+      const metadata = this.getMetadata(original, args, span);
+
+      const setter: types.HeaderSetter = {
+        setHeader(name: string, value: string) {
+          metadata.set(name, value);
+        }
+      };
+
+      const propagation = plugin.tracer.propagation;
+      if (propagation) {
+        propagation.inject(setter, span.spanContext);
+      }
+
+      span.addAttribute(GrpcPlugin.ATTRIBUTE_GRPC_METHOD, original.path);
+      span.addAttribute(GrpcPlugin.ATTRIBUTE_GRPC_KIND, span.type);
+
+      const call = original.apply(self, args);
+
+      plugin.tracer.wrapEmitter(call);
+
+      // if server stream or bidi
+      if (original.responseStream) {
+        let spanEnded = false;
```

Comment:
Can you add a comment saying that `spanEnded` is needed because it's possible that both `error` and `status` events are emitted?